golang/cmd/event-ingest/main.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/twmb/franz-go/pkg/kgo"
+)
+
+const batchSize = 1000
+
+type Event struct {
+	OrganizationID int                    `json:"organization_id,omitempty"`
+	CustomerID     *int                   `json:"customer_id,omitempty"`
+	CustID         string                 `json:"cust_id,omitempty"`
+	EventName      string                 `json:"event_name,omitempty"`
+	TimeCreated    time.Time              `json:"time_created,omitempty"`
+	Properties     map[string]interface{} `json:"properties,omitempty"`
+	IdempotencyID  string                 `json:"idempotency_id,omitempty"`
+	InsertedAt     time.Time              `json:"inserted_at,omitempty"`
+}
+
+func main() {
+	seeds := []string{"localhost:9092"}
+	ctx := context.Background()
+
+	// Setup kafka consumer
+	cl, err := kgo.NewClient(
+		kgo.SeedBrokers(seeds...),
+		kgo.ConsumerGroup("default"),
+		kgo.ConsumeTopics("test-topic"),
+		kgo.DisableAutoCommit(),
+	)
+	if err != nil {
+		panic(err)
+	}
+	defer cl.Close()
+
+	//setup db connection
+	db, err := sql.Open("postgres", "postgres://user:password@localhost/dbname?sslmode=disable")
+	if err != nil {
+		panic(err)
+	}
+	defer db.Close()
+
+	// Set up prepared statement for insert
+	stmt, err := db.Prepare(`INSERT INTO events (organization_id, customer_id, cust_id, event_name, time_created, properties, idempotency_id, inserted_at)
+		VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
+		ON CONFLICT DO NOTHING`)
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		fetches := cl.PollFetches(ctx)
+		if fetches == nil {
+			continue
+		}
+		if fetches.IsClientClosed() {
+			return
+		}
+		if errs := fetches.Errors(); len(errs) > 0 {
+			// All errors are retried internally when fetching, but non-retriable errors are
+			// returned from polls so that users can notice and take action.
+			panic(fmt.Sprint(errs))
+		}
+
+		events := make([]*Event, 0)
+		fetches.EachRecord(func(r *kgo.Record) {
+			event := &Event{}
+			err := json.Unmarshal(r.Value, event)
+			if err != nil {
+				fmt.Printf("Error unmarshalling event: %s\n", err)
+				return
+			}
+			event.CustomerID = nil
+			event.InsertedAt = time.Now()
+			events = append(events, event)
+
+			if len(events) >= batchSize {
+				err = insertEvents(stmt, events, db, ctx)
+				if err != nil {
+					fmt.Printf("Error inserting events into database: %s\n", err)
```

Comment:
As noted above (and below), you probably want to do the individual inserts before this check, and in this check all you need to do is a `tx.Commit()`.

However, if the commit fails, you want to panic here - because if you don't, you'll lose data (the Kafka records will be committed). As long as your ingress table is simple, the only reason you'll get an error here is if something really bad happens.

Reply:
Wouldn't a simple return work better than a panic? that way we can keep running. That's what we do on the other one